### PR TITLE
Shared memory support

### PR DESCRIPTION
**Pull request by atilag**
I was trying to add Shared Memory support in the wasmtime-py library, but I found out that C-API support here was not finished, so this is my bold attempt to implement something that works.

As far as a I read, threads on WASI are still in discussion and Shared Memory is part of the proposal. In other hand there's some basic support for Shared memory already present in the Rust API interface, so I guess it's fine to try gluing the C-API with what is supported today, and make wasmtime-py works with Shared Memory (?).

### CODE REVIEW

File: crates/fuzzing/src/oracles/dummy.rs
Comment (atilag) [inline crates/fuzzing/src/oracles/dummy.rs:26]: Ok, this is wrong.

File: crates/c-api/src/extern.rs
Comment (alexcrichton) [inline crates/c-api/src/extern.rs:23]: There's a split between the `wasm_*` APIs and `wasmtime_*` APIs and we can't change the `wasm_*` ones so this unfortunately can't change. Perhaps rename the `todo!()` though to a `panic!()` with an error message along the lines of "please use the `wasmtime_*` method instead"?

I should also note that if the `wasmtime_*` method doesn't exist yet and you need this method specifically then it's fine to add a `wasmtime_*` method too.

File: crates/c-api/src/types/extern.rs
Comment (alexcrichton) [inline crates/c-api/src/types/extern.rs:18]: This I think we'll want to avoid where we decided to split the runtime representation of `Memory` and `SharedMemory` but at the type level they're the same type. In that sense the `Memory` variant above this should be sufficient.

File: crates/c-api/src/types/memory.rs
Comment (alexcrichton) [inline crates/c-api/src/types/memory.rs:83]: Mind panicking here for now with `unimplemented!()` if both `shared` and `memory64` are set? (we need to add a constructor for that but haven't gotten around to it yet)

Comment (alexcrichton) [reply crates/c-api/src/types/memory.rs:83]: Actually I take this back, this may want to use `MemoryTypeBuilder` to support shared 64-bit memories

**Comment by atilag** [comment]
Suggestions applied!
Wanted to link this issue with [this other one] from wasmtime-py, as this is a required dependency. We will need to rebuild the `_bindings.py` after this is merged.

**Comment by alexcrichton** [comment]
Thanks! I think we'll still want to remove the `SharedMemory` arm of `ExternType` which will help simplify this a bit more though.

**Comment by atilag** [comment]
Done!

File: crates/fuzzing/src/oracles/dummy.rs
Comment (alexcrichton) [inline crates/fuzzing/src/oracles/dummy.rs:83]: Is this still needed? (or should it perhaps be integrated elsewhere?)

**Review by alexcrichton — COMMENTED** [review]
Looking good! With `MemoryTypeBuilder` for the implementation I think this will be good to go 👍

File: crates/fuzzing/src/oracles/dummy.rs
Comment (atilag) [reply crates/fuzzing/src/oracles/dummy.rs]: Nah, I just removed it

**Comment by atilag** [comment]
Oh, nice, the MemoryTypeBuilder made the code way more readable :)
Done!

**Review by alexcrichton — APPROVED** [review]
There's a minor issue with clang-format in CI, but otherwise I think this is good to go 👍

**Comment by atilag** [comment]
Oh, no problem. I can run clang-format and see a green CI :)

**Comment by alexcrichton** [comment]
I think there might be something else with clang-format in CI?

If you'd like also feel free to mark this as "ready for review" so it can be approved+merged once CI passes.

**Comment by atilag** [comment]
Yeah, might be the version of clang format usedd in the CI (15?) has different default options than the one I use locally.
Anyway, ready!
Thanks for taking a look Alex!

**Comment by alexcrichton** [comment]
I dunno what's going on with clang-format myself, I've attempted to resolve the latest issue with a new commit

**Pull request merged**
